File: mnist-ex/mnist-par.R
# Load the packages
library(farff)
library(OpenML)
library(randomForest)
library(dplyr)
library(ggplot2)
library(caret)

# Load the Fashion MNIST dataset
fashion_mnist <- getOMLDataSet(data.id = 40996)

# Extract and prepare data
X <- as.data.frame(fashion_mnist$data)
# Check the names of the columns to identify the target column
print(names(X))

# Assuming 'class' is the name of the target column
y <- as.factor(X$class)  
X <- X[, names(X) != "class"]  # Remove the label column from X

# Split dataset into training and testing
set.seed(42)
train_indices <- sample(1:nrow(X), 0.7 * nrow(X))
X_train <- X[train_indices, ]
y_train <- y[train_indices]
X_test <- X[-train_indices, ]
y_test <- y[-train_indices]

# Train the Random Forest model
start_time <- Sys.time()
rf_model <- randomForest(x = X_train, y = y_train, ntree = 100, mtry = 28, do.trace = 50, njobs = 10)
end_time <- Sys.time()
training_time <- end_time - start_time
print(paste("Training time: ", training_time))

# Predict on the test set
predictions <- predict(rf_model, X_test)

# Calculate accuracy and other metrics
accuracy <- mean(predictions == y_test)
print(paste("Accuracy:", accuracy))
confusion_matrix <- table(Predicted = predictions, Actual = y_test)
print(confusion_matrix)
classification_report <- confusionMatrix(predictions, y_test)
print(classification_report)

File: mnist-ex/mnist-par.py
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report, confusion_matrix
from sklearn.datasets import fetch_openml
import time

# Load the Fashion MNIST dataset from OpenML
X, y = fetch_openml('Fashion-MNIST', version=1, return_X_y=True)

# Split dataset into training and testing
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Initialize the Random Forest Classifier with n_jobs=10
clf = RandomForestClassifier(random_state=42, n_jobs=10)

# Start the timer
start_time = time.time()

# Train the model
clf.fit(X_train, y_train)

# End the timer
end_time = time.time()

# Calculate the training time
training_time = end_time - start_time
print(f"Training Time: {training_time} seconds")

# Predict on the test set
y_pred = clf.predict(X_test)

# Evaluate the model
accuracy = clf.score(X_test, y_test)
print(f"Model Accuracy: {accuracy}")

# Confusion matrix
print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred))

# Classification report
print("\nClassification Report:")
print(classification_report(y_test, y_pred))
